File: rest-common/src/main/java/org/chronopolis/rest/support/ZonedDateTimeDeserializer.java
package org.chronopolis.rest.support;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

/**
 * Gson deserializer for zoned date time from an ISO-8601 string
 *
 * Created by shake on 3/22/16.
 */
public class ZonedDateTimeDeserializer implements JsonDeserializer<ZonedDateTime> {
    @Override
    public ZonedDateTime deserialize(JsonElement jsonElement, Type type, JsonDeserializationContext jsonDeserializationContext) throws JsonParseException {
        DateTimeFormatter fmt = DateTimeFormatter.ISO_OFFSET_DATE_TIME.withZone(ZoneOffset.UTC);
        return ZonedDateTime.from(fmt.parse(jsonElement.getAsString()));
    }
}

File: rest-models/src/main/kotlin/org/chronopolis/rest/models/serializers/ZonedDateTimeDeserializer.kt
package org.chronopolis.rest.models.serializers

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonDeserializer
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter.ISO_OFFSET_DATE_TIME

/**
 * Simple deserializer for reading [ZonedDateTime] from an ISO-8601 string at UTC-0
 *
 * @author shake
 */
class ZonedDateTimeDeserializer : JsonDeserializer<ZonedDateTime>() {

    override fun deserialize(jsonParser: JsonParser,
                             deserializationContext: DeserializationContext): ZonedDateTime {
        val text = jsonParser.text
        return ZonedDateTime.from(ISO_OFFSET_DATE_TIME.withZone(ZoneOffset.UTC).parse(text))
    }
}

